Mathlib/Algebra/Divisibility/Ring.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Divisibility.Basic
+import Mathlib.Data.Nat.Choose.Sum
+import Mathlib.Tactic.Linarith
+
+/-!
+# Lemmas about divisibility in rings
+
+# Main results:
+ * `dvd_smul_of_dvd`: stating that `x ∣ y → x ∣ m • y` for any scalar `m`.
+ * `Commute.pow_dvd_add_pow_of_pow_eq_zero_right`: stating that if `y` is nilpotent then
+   `x ^ m ∣ (x + y) ^ p` for sufficiently large `p` (together with many variations for convenience).
+-/
+
+variable {R : Type*} {x y : R}
+
+lemma dvd_smul_of_dvd {M : Type*} [SMul M R] [Semiring R] [SMulCommClass M R R]
```

Comment:
Do we really need `Semiring` here? Maybe some nonunital nonassociative thing will work?

Reply:
Good point, this can be `Semigroup R` instead of `Semiring R`. Associativity is necessary for the reason explained here https://github.com/leanprover-community/mathlib4/pull/7355#discussion_r1335240589